### PR TITLE
Add and update specs for IO methods

### PR DESCRIPTION
**Pull request by javanthropus**
This PR adds missing specs, updates error message expectations, refactors redundant test fixtures, and adds coverage for edge cases and version-specific behaviors across `core/io`.

### Summary of Changes

* **`IO#binmode`**: Added spec verifying that newline conversion is disabled when reading in binary mode.
* **`IO#getbyte`**, **`IO#read`**, **`IO#readbyte`**, **`IO#readpartial`**: Added specs for reading after `ungetc`, including specs verifying that byte-oriented methods raise `IOError` ("byte oriented read for character buffered IO") when character conversion is enabled.
* **`IO#internal_encoding`**: Added specs for `BINARY` external encoding, including Ruby 3.3+ behavior where internal encoding assignments are ignored.
* **`IO#pwrite`**: Added spec verifying `Errno::EINVAL` is raised on negative offsets.
* **`IO#read_nonblock`**: Added spec for CRuby Bug [#18421](https://bugs.ruby-lang.org/issues/18421) (fixed in 3.0.4) ensuring 0-length reads clear the buffer, and updated `IOError` expectation.
* **`IO#reopen`**: Added specs for excess argument validation and mode preservation (`r`, `w`, `r+`, `a`) when reopening with a String path.
* **`IO#seek`**: Added spec for CRuby Bug [#20919](https://bugs.ruby-lang.org/issues/20919) (fixed in 3.4) verifying character buffer is cleared after `ungetc`.
* **`IO#set_encoding`**: Refactored `set_encoding nil, nil` mode blocks to reuse shared behaviors. Added specs for `"-"` internal encoding, ASCII compatibility validation, newline options, ASCII-incompatible encodings, and Ruby 3.2+/3.3+ version specific behaviors.
* **`IO#sysread`**, **`IO#sysseek`**: Updated `IOError` expectation messages for buffered IO and added specs for non-empty character buffers.
* **`IO#syswrite`**: Refactored pipe test setup and added specs for non-blocking write behavior on full pipes (raising `Errno::EAGAIN` / `Errno::EWOULDBLOCK` on Ruby < 3.2).
* **`IO#write`**: Added specs for writing binary data with `ASCII-8BIT` external encoding and newline conversion.

### Verification

Ran specs via `mspec core/io/`:
`103 files, 1582 examples, 2459 expectations, 0 failures, 0 errors, 0 tagged`

### CODE REVIEW

File: core/io/binmode_spec.rb [hunk 1]
```diff
@@ -31,6 +31,19 @@
...
+
+    @io = new_io(@name, "wb")
+    @io.write(data)
+    @io.close
+
+    @io = new_io(@name, "rt")
+    @io.set_encoding("utf-8:ISO-8859-1", newline: :universal)
+    @io.binmode
+    @io.read.should == data
+  end
```
Comment (andrykonchin) [inline core/io/binmode_spec.rb:46]: **suggestion:** `#binmode` affects not only the `#read` method but also some other "reading" methods - each, each_line, each_char, each_codepoint, getc, gets, readchar, readline, readlines.

Some "readable" methods aren't affected - e.g. sysread, pread, read_nonblock, readpartial, readbyte, each_byte).

File: core/io/getbyte_spec.rb [hunk 2]
```diff
@@ -25,6 +25,22 @@
...
+  end
+
+  it "raises an exception after ungetc with character conversion" do
+    @io.set_encoding("utf-8:utf-16be")
+    c = @io.getc
+    @io.ungetc(c)
+    -> do
+      @io.getbyte
+    end.should.raise(IOError, "byte oriented read for character buffered IO")
+  end
```
Comment (andrykonchin) [inline core/io/getbyte_spec.rb]: **minor:** this appears to be a CRuby-specific implementation detail (or an intentional design constraint) that other Ruby implementations, such as JRuby, TruffleRuby, or Natalie, might handle differently (e.g., by
permitting byte-oriented reading under these conditions).

CRuby already allows mixing byte-oriented and character-oriented read operations on streams with multibyte encodings (such as UTF-16BE). Semantically, calling `ungetc` does not make this behavior any worse or introduce additional issues.

File: core/io/internal_encoding_spec.rb [hunk 3]
```diff
@@ -93,6 +93,27 @@
...
+      @io.internal_encoding.should == nil
+    end
+
+    ruby_version_is ""..."3.3" do
+      it "returns the value set by #set_encoding when the external encoding is BINARY" do
+        @io = new_io @name, @object
+        @io.set_encoding(Encoding::BINARY, Encoding::IBM437)
+        @io.internal_encoding.should.equal?(Encoding::IBM437)
+      end
+    end
```
Comment (andrykonchin) [inline core/io/internal_encoding_spec.rb]: **minor:** at present, `ruby/spec` supports Ruby versions starting from 3.3. Consequently, the test case above is redundant.

The `ruby_version_is` guard bellow is not needed as well.

File: core/io/internal_encoding_spec.rb [hunk 4]
```diff
@@ -93,6 +93,27 @@
...
+
+    ruby_version_is "3.3" do
+      it "ignores the value set by #set_encoding when the external encoding is BINARY and returns nil" do
+        @io = new_io @name, @object
+        @io.set_encoding(Encoding::BINARY, Encoding::IBM437)
+        @io.internal_encoding.should == nil
+      end
+    end
```
Comment (andrykonchin) [inline core/io/internal_encoding_spec.rb]: **suggestion:** similar cases are tested in `set_encoding_spec.rb` so it may be a better place and for this one.

File: core/io/pwrite_spec.rb [hunk 5]
```diff
@@ -59,10 +59,12 @@
...
-  it "raises a Errno::EINVAL if the offset is invalid" do
-    -> {
-      @file.pwrite("foo", -1)
-    }.should.raise(Errno::EINVAL)
+  platform_is_not :windows do
+    it "raises a Errno::EINVAL if the offset is invalid" do
+      -> {
+        @file.pwrite("foo", -1)
+      }.should.raise(Errno::EINVAL)
+    end
```
Comment (andrykonchin) [inline core/io/pwrite_spec.rb]: **suggestion:** Actually it's interesting what happens on Windows when given a negative value. Yeah, here we test POSIX/Windows system calls but it 's still a part of the CRuby public behavior.

**Review by andrykonchin — COMMENTED** [review]
Will continue review later.

File: core/io/binmode_spec.rb (hunk 1 above)
Comment (javanthropus) [reply core/io/binmode_spec.rb:46]: Are you asking for additional tests for each of those cases?

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (javanthropus) [reply core/io/getbyte_spec.rb]: My understanding is that this project documents the behavior of CRuby, and this is indeed a CRuby behavior.  I've not tried it on other implementations.  Is that a requirement now?

File: core/io/internal_encoding_spec.rb (hunk 3 above)
Comment (javanthropus) [reply core/io/internal_encoding_spec.rb]: I'll clean this up.

File: core/io/internal_encoding_spec.rb (hunk 4 above)
Comment (javanthropus) [reply core/io/internal_encoding_spec.rb]: Either place works, but is there a rule to help decide which is best?

File: core/io/pwrite_spec.rb (hunk 5 above)
Comment (javanthropus) [reply core/io/pwrite_spec.rb]: I had to add the guard here after my first attempt broke the tests on Windows.  I don't know what the right behavior _should be_ on Windows, so I just left it out.  This may also have different behaviors on different Ruby implementations.

**Review by javanthropus — COMMENTED** [review]
Thanks for taking a look, @andrykonchin.

File: core/io/binmode_spec.rb (hunk 1 above)
Comment (andrykonchin) [reply core/io/binmode_spec.rb:46]: I'm not asking for them as a requirement, but if you have the time and interest, extra tests for those methods would be very appreciated.

File: core/io/internal_encoding_spec.rb (hunk 4 above)
Comment (andrykonchin) [reply core/io/internal_encoding_spec.rb]: My rule of thumb is simply: Where is the most intuitive place for a developer to look for them first?

File: core/io/pwrite_spec.rb (hunk 5 above)
Comment (andrykonchin) [reply core/io/pwrite_spec.rb]: `pwrite` raises `Errno::EINVAL` for offsets of -3 and below, but simply writes to the file at the current position when the offset is -1 or -2. Replacing -1 with -100 should make this test work on Windows as well.

I suspect this special handling of -1 and -2 is unintentional, so it probably doesn't make sense to add test cases for them.

File: core/io/set_encoding_spec.rb [hunk 6]
```diff
@@ -26,23 +115,28 @@
...
-    Encoding.default_external = Encoding::IBM437
-    Encoding.default_internal = Encoding::IBM866
+    Encoding.default_external = Encoding::UTF_8
+    Encoding.default_internal = nil
+
+    @io.set_encoding nil, nil
 
     @io.external_encoding.should == nil
     @io.internal_encoding.should == nil
   end
```
Comment (andrykonchin) [inline core/io/set_encoding_spec.rb]: Not sure if understand the reason of the changes in the test above.

File: core/io/binmode_spec.rb (hunk 1 above)
Comment (javanthropus) [reply core/io/binmode_spec.rb:46]: Sure.  That should be easy enough.

File: core/io/internal_encoding_spec.rb (hunk 4 above)
Comment (javanthropus) [reply core/io/internal_encoding_spec.rb]: It's easy enough to move this over.  It's similar enough to the other set_encoding tests.

File: core/io/pwrite_spec.rb (hunk 5 above)
Comment (javanthropus) [reply core/io/pwrite_spec.rb]: I'll give that a shot.  I feel like I should also open a bug against CRuby about the -1 and -2 cases.

File: core/io/set_encoding_spec.rb (hunk 6 above)
Comment (javanthropus) [reply core/io/set_encoding_spec.rb]: Good catch.  I've had these edits on the back burner for so long, it's hard for me to be sure what's going on with this change.  It looks like 2 specs were chopped up and merged together.  I'll look into this.

Comment (javanthropus) [reply core/io/set_encoding_spec.rb]: This should be cleaned up now.

**Review by andrykonchin — APPROVED** [review]

**Comment by andrykonchin** [comment]
Thank you!

**Comment by javanthropus** [comment]
Thanks for the review and approval, @andrykonchin.  When will this be merged?

**Comment by andrykonchin** [comment]
Hi, thanks for following up! A few items need a quick maintainer discussion before we can merge. We'll update you shortly!

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (eregon) [reply core/io/getbyte_spec.rb]: > My understanding is that this project documents the behavior of CRuby

Not exactly, as the README says:
> The Ruby Spec Suite, abbreviated ruby/spec, is a test suite for the behavior of the Ruby programming language.

Which notably means avoiding to add specs for things that are bugs in CRuby (and instead report the bug/inconsistency), ask clarification on the CRuby bug tracker for unclear cases and otherwise yes mostly assume CRuby behavior is intended if it looks reasonable.

I think this is a case of "unclear, should ask clarification on the CRuby bug tracker".
Would you mind opening a Misc issue at https://bugs.ruby-lang.org/ to ask the reason for this error and detail there how it's inconsistent?

I think for now it's better to not add these specs in ruby/spec, because the effect would be probably unclear semantics and extra overhead for other Ruby implementations, when that behavior might be undesirable anyway (could probably be even seen as a bug by someone trying to do legitimately).
Could you remove them for this PR?

File: core/io/pwrite_spec.rb (hunk 5 above)
Comment (eregon) [reply core/io/pwrite_spec.rb]: I suspect -1/-2 are probably the underlying libc/syscall function accepting those.
Not sure how much we can spec there.
But yes, could be worth a CRuby issue to ask if a reliable exception would be better.

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (eregon) [reply core/io/getbyte_spec.rb]: Playing a bit in IRB with this and yeah it's weird, I could maybe understand if byte operations are not allowed with UTF-16/32 (which are the only encodings with min-length > 1 byte), although that would be inconsistent with String which doesn't mind.
But it's not even that, that's allowed, it's indeed after ungetc it's not allowed.
That starts to feel like a bug.

Comment (eregon) [reply core/io/getbyte_spec.rb]: Ah BTW there was already 1 spec about this (`it "raises an exception after ungetc with data in the buffer and character conversion enabled" do`) let's remove it too

File: core/io/write_spec.rb [hunk 7]
```diff
@@ -140,6 +140,36 @@
...
   end
+
+  it "writes binary data if encoding is ASCII-8BIT" do
+    File.open(@filename, "w:ascii-8bit") do |file|
+      file.write('Hëllö'.encode('ISO-8859-1'))
+    end
+    ë = ([235].pack('U')).encode('ISO-8859-1')
+    ö = ([246].pack('U')).encode('ISO-8859-1')
+    res = "H#{ë}ll#{ö}"
+    File.binread(@filename).should == res.force_encoding(Encoding::BINARY)
```
Comment (eregon) [inline core/io/write_spec.rb]: `res.b` is a shorter form for `.force_encoding(Encoding::BINARY)`, could you use it?

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (javanthropus) [reply core/io/getbyte_spec.rb]: I will open tickets as requested, but I've had mixed results getting attention on some of them.  This [ticket](https://bugs.ruby-lang.org/issues/18995) is one such which is being ignored.  Is there something else I should do to avoid these new tickets also being ignored?

**Review by eregon — COMMENTED** [review]
Thank you, looks great overall.
As discussed https://github.com/ruby/spec/pull/1384#discussion_r3677892343 I think we shouldn't add specs for `byte oriented read for character buffered IO` IOError since that seems inconsistent and very low value compatibility to do the same on other Ruby implementations.

Could you remove them?
You could do so in a separate commit so if we'd ever change our mind on this we can easily restore those.

(there is some back-and-forth in these commits, but the commit messages are helpful and it's nice for review so I plan to merge with rebase, i.e. not squashing)

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (javanthropus) [reply core/io/getbyte_spec.rb]: I was looking over the implementation, and I think I understand the reasoning for rejecting byte oriented reads in these cases.  If such reads are allowed, you have to answer the question of which bytes to return.

Do you return the buffered bytes first?  If so, do you return them as-is, or do you attempt to undo the effective character conversion that is assumed to have been performed on them?  If you ignore the buffered bytes, do you clear the buffer or just ignore it and return the next unbuffered byte?

It's clear to me that the CRuby maintainers chose the safest choice by raising an error in this case.  Answering the above questions is a matter of opinion as to how to handle what is arguably a bad implementation decision of mixing character oriented and byte oriented read operations.  Where behavior is unambiguous, Ruby gets out of the way, but where the "right" solution is indeterminate, it stops the code as an alert to the author to rethink their solution.

I can still open a ticket, but I have several open tickets with similar questions which have all been ignored for a year or more.  This ticket will be a similarly difficult one to answer, and I don't think it will get any traction unless I can get the attention of a core maintainer.

In the meantime, I will remove these as requested so that the commit that removes them can be easily reverted in the future if/when a decision is made about the behaviors under test.

File: core/io/write_spec.rb (hunk 7 above)
Comment (javanthropus) [reply core/io/write_spec.rb]: I've done this for my additions, but note that there are several other places in this file that use force_encoding in this way which I did not modify since they are not part of this MR.

File: core/io/read_nonblock_spec.rb [hunk 8]
```diff
@@ -66,14 +66,16 @@
...
     @read.set_encoding(
       'utf-8', universal_newline: true
     )
     c = @read.getc
     @read.ungetc(c)
-    -> { @read.read_nonblock(3).should == "foo" }.should.raise(IOError)
+    -> do
+      @read.read_nonblock(3)
+    end.should.raise(IOError, "byte oriented read for character buffered IO")
   end
```
Comment (javanthropus) [inline core/io/read_nonblock_spec.rb]: @eregon, while removing the tests as you requested, this one caught my eye because it's a pre-existing test that does exactly the same thing.  All I did here was clean it up and add a check for the error message in addition to the exception type.

In light of this test having already been accepted [back in 2022](https://github.com/ruby/spec/commit/787577f4ef), maybe it's safe to say that other Ruby implementations already have this logic correctly implemented.  If you still think that this class of test should be removed, then I think it best to remove this one as well.  What do you think?

File: core/io/sysread_spec.rb [hunk 9]
```diff
@@ -52,7 +52,17 @@
...
+    end.should.raise(IOError, "sysread for buffered IO")
+  end
+
+  it "raises an error when called with a non-empty character buffer" do
+    @file.set_encoding(Encoding::UTF_8, Encoding::UTF_16BE)
+    @file.ungetc("a".encode(Encoding::UTF_16BE))
+    -> do
+      @file.sysread(5)
+    end.should.raise(IOError, "byte oriented read for character buffered IO")
   end
```
Comment (javanthropus) [inline core/io/sysread_spec.rb]: @eregon, this test seems to be the same class as the others you want removed for now.  It feels lower level to me though, so please confirm if this should also be removed.

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (eregon) [reply core/io/getbyte_spec.rb]: > Do you return the buffered bytes first?

I would think so.

> If so, do you return them as-is, or do you attempt to undo the effective character conversion that is assumed to have been performed on them?

That part I don't know so well, when one does `ungetc` is there any encoding conversion?
For `ungetbyte` I assume there isn't.
I'd think for `ungetc` it should ideally be just the same as a few `ungetbyte` in the end (and possibly some conversion before that), and so `readbyte` after would just return the bytes put in the buffer.

Maybe we could look at how StringIO behaves, I'm guessing it doesn't raise an error in that case.

Comment (eregon) [reply core/io/getbyte_spec.rb]: > I will open tickets as requested, but I've had mixed results getting attention on some of them. This [ticket](https://bugs.ruby-lang.org/issues/18995) is one such which is being ignored. Is there something else I should do to avoid these new tickets also being ignored?

Yes sometimes people answer and sometimes they don't, maybe because nobody knows or remember.
One possibility is to add a ticket to a DevMeeting, then usually someone will take a look at it.
The next dev meeting isn't scheduled yet though.

File: core/io/read_nonblock_spec.rb (hunk 8 above)
Comment (eregon) [reply core/io/read_nonblock_spec.rb]: I'd remove this one too, like in https://github.com/ruby/spec/pull/1384#discussion_r3677960286.
If it's many of them then yeah let's reconsider but if a handful let's just remove those.
If you miss some it's also not a big deal.

> maybe it's safe to say that other Ruby implementations already have this logic correctly implemented

TruffleRuby doesn't have it, Rubinius doesn't have it.
JRuby seems to have some of it due to copying that code from CRuby, but I don't think they have given it much thought.

File: core/io/sysread_spec.rb (hunk 9 above)
Comment (eregon) [reply core/io/sysread_spec.rb]: Yes I think anything that checks for a `byte oriented read for character buffered IO` IOError shouldn't be spec'd at this point.

File: core/io/getbyte_spec.rb (hunk 2 above)
Comment (eregon) [reply core/io/getbyte_spec.rb]: Looking at `READ_CHAR_PENDING`/`READ_DATA_PENDING` in CRuby and them having different buffers seems very much like accidental complexity to me, not something we'd want to spec, but some implementation artifact from old Ruby versions. The Ruby IO API is also known as one of the most messy.

Comment (eregon) [reply core/io/getbyte_spec.rb]: Unpopular opinion but maybe ungetc/ungetbyte should be deprecated/removed, it adds significant complexity for little value and little need. I'm not hopeful though.

Read buffering itself is fine and needed e.g. for `IO#readline`, but being able to mutate the buffer though `ungetc/ungetbyte` that seems quite brittle.

Comment (javanthropus) [reply core/io/getbyte_spec.rb]: > That part I don't know so well, when one does ungetc is there any encoding conversion?

No, the bytes passed to ungetc are dropped directly into the internal character buffer, which is separate from the internal byte buffer.  The idea is that you get those bytes back when calling getc or any other character oriented read operation.  You can even intentionally put an invalid character back via ungetc, so it's not even possible to always reverse the character conversion of the stream in order to drop the bytes into the byte buffer.

In order to attempt a reverse conversion, the behavior of ungetc would have to change to raise a conversion error when the conversion fails.  It never does that now, so that would risk breaking existing code in surprising ways.  Then there's the cost of creating a new character convertor for the reverse operation and risk of state loss when resetting the state of the forward character convertor after pushing converted bytes back into the byte buffer.  This quickly becomes a can of worms.

I think we're approaching the point of identifying these specs as undefined behavior unless the Ruby community sees value in aligning on an implementation.

Comment (javanthropus) [reply core/io/getbyte_spec.rb]: > Looking at READ_CHAR_PENDING/READ_DATA_PENDING in CRuby and them having different buffers seems very much like accidental complexity to me, not something we'd want to spec, but some implementation artifact from old Ruby versions.

I think it's deliberate complexity given my earlier message.  It provides the least surprise in the more common case where people don't mix character and byte oriented operations on the same IO object.

> The Ruby IO API is also known as one of the most messy.

I'm far too aware of that mess.  I've spent years of effort attempting to make a faithful, pure Ruby implementation of IO operations in the [io-like gem](https://rubygems.org/gems/io-like).  I've been successful, which is why I have all these new specs for weird corner cases.

> Unpopular opinion but maybe ungetc/ungetbyte should be deprecated/removed, it adds significant complexity for little value and little need.

People implementing parsers at least would probably disagree with you.  It's common to leverage the read buffer as a form of lookahead by putting back bytes or characters that you just peeked at in order to make a decision.

> being able to mutate the buffer though ungetc/ungetbyte that seems quite brittle.

Whether brittle or not, I used ungetc in these tests as a way to ensure that the character buffer has predictable content.  There may be other ways, but this was the easiest to prove and use.  The point is that the character buffer may have content in it.  That content is derived from the byte buffer, so bytes were consumed to provide the character buffer content.

The "right" solution for a byte oriented read operation in that condition is debatable, but it's hard to imagine the CRuby maintainers changing their decision at this point.  Nothing less than a complete overhaul of the IO class would justify that sort of change.

Comment (eregon) [reply core/io/getbyte_spec.rb]: > Maybe we could look at how StringIO behaves, I'm guessing it doesn't raise an error in that case.

Indeed, seems StringIO does not mind: https://github.com/ruby/stringio/blob/master/ext/stringio/stringio.c
So maybe IO should align to StringIO on CRuby then.

> You can even intentionally put an invalid character back via ungetc, so it's not even possible to always reverse the character conversion of the stream in order to drop the bytes into the byte buffer.

If one is gonna put an invalid character, I think might as well use `ungetbyte`.

Cases of actual encoding conversion on IO are already very rare, I personally don't see a good enough reason to e.g. support broken characters for `ungetc`, raising there would be more sensible and simplify things significantly (no need for an extra character buffer).

I also imagine it's pretty easy to build your own ungetc-like functionality on top of an IO, and it's something that feels like it doesn't belong to IO, unless it trivially extends the existing buffering, but that isn't the case.

This is of course my opinion, but this area seems messy enough and low value enough for users that I don't think we want to enforce the behavior, at least not until there is actual code relying on this for good reasons / reasonable use case.
Another POV is CRuby devs sometimes are unhappy about specifying too much details including things which might be accidental, this certainly feels like it fits that bill.

Comment (javanthropus) [reply core/io/getbyte_spec.rb]: > Indeed, seems StringIO does not mind

StringIO diverges from IO's behavior in many places.  I'm not sure it's a good example of what to do, at least not at this time.

> If one is gonna put an invalid character, I think might as well use `ungetbyte`.

The point is that you can put an invalid character back into the character buffer via `ungetc` such that there's no way to reverse the conversion to the byte buffer.  If you want to eliminate the character buffer, then you either have to reverse the encoded character so that followup byte oriented read operations get the expected bytes or error out when you can't perform the conversion for any reason.

That sounds quite reasonable to me, but I feel like that kind of change will have to wait for Ruby 5.0 because the current CRuby behavior is fully consistent everywhere I've tested and has been so for all versions since 2.7.

> I also imagine it's pretty easy to build your own ungetc-like functionality on top of an IO, and it's something that feels like it doesn't belong to IO, unless it trivially extends the existing buffering, but that isn't the case.

I've done that in the io-like gem.  It's honestly not that easy when you want all of the read operations to work transparently with it.  You have to hack each of the read methods to account for whatever standalone buffer you use.  Either that or you have to deal with the reverse conversion problem I mentioned so you can convert the `ungetc` call to an equivalent `ungetbyte` call.  A limited IO wrapper class could be minted just to expose the necessary behaviors for a specific workload, but I digress.

Anyway, I'm not here to argue the merits of `ungetc` and `ungetbyte` or their current behaviors.  I just need to figure out if the behavior is something that should be part of the specification or left unspecified.  As you said, the CRuby devs may not want this behavior locked down.  I'll have to open the ticket and hope it gets an answer.

Thank you for all the discussion on this.

Comment (eregon) [reply core/io/getbyte_spec.rb]: >  If you want to eliminate the character buffer, then you either have to reverse the encoded character so that followup byte oriented read operations get the expected bytes or error out when you can't perform the conversion for any reason.

Yes, I think this would be the best.

**Review by eregon — COMMENTED** [review]
Could you remove all specs which depend on the `byte oriented read for character buffered IO` behavior in its own commit?
Then happy to merge this :)

**Comment by javanthropus** [comment]
I opened https://bugs.ruby-lang.org/issues/22230 for discussion regarding raising IOError during byte oriented reads when the character buffer is nonempty and then removed the questionable tests in a single, easy to revert commit.
